README.md
# ![ÖbEngine](https://raw.githubusercontent.com/ObEngine/ObWiki/master/banner.png)

[![Discord](https://img.shields.io/discord/302123079818149888.svg?style=for-the-badge&logo=discord)](https://discordapp.com/invite/HkM6zdT)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=for-the-badge)](https://raw.githubusercontent.com/ObEngine/ObEngine/master/license)
[![Github All Releases](https://img.shields.io/github/downloads/sygmei/obengine/total.svg?style=for-the-badge)](https://github.com/ObEngine/ObEngine/releases)
[![Codacy Badge](https://img.shields.io/codacy/grade/81846da629d244f69f357568a2f8d690.svg?style=for-the-badge)](https://www.codacy.com/app/Sygmei/ObEngine?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Sygmei/ObEngine&amp;utm_campaign=Badge_Grade)

## CI/CD status

| CI/CD name  | Build status |
| ----------- | ------------ |
| GitHub Actions | [![Build Status](https://img.shields.io/github/workflow/status/ObEngine/ObEngine/build?logo=github&style=for-the-badge)](https://github.com/ObEngine/ObEngine/actions?query=workflow%3Abuild) |
| Travis      | [![Build Status](https://img.shields.io/travis/Sygmei/ObEngine.svg?style=for-the-badge&logo=travis)](https://travis-ci.org/Sygmei/ObEngine) |
| AppVeyor    | [![Build status](https://img.shields.io/appveyor/ci/Sygmei/ObEngine.svg?style=for-the-badge&logo=appveyor)](https://ci.appveyor.com/project/Sygmei/obengine) |
| Azure Pipelines | [![Build Status](https://img.shields.io/azure-devops/build/obengine/d879c6ea-1ed9-4984-b3d1-84c5dd9b8feb/1.svg?style=for-the-badge&logo=microsoft-azure)](https://dev.azure.com/obengine/%C3%96bEngine) |

## Useful links

- :book: [ÖbEngine's Wiki](https://github.com/ObEngine/ObEngine/wiki)
- :link: [ÖbEngine's Website](https://www.obengine.io)
- :page_facing_up: [ÖbEngine's Documentation](https://www.obengine.io/doc.html)
- :speech_balloon: [ÖbEngine's Discord](https://discord.gg/HkM6zdT)
- :hammer_and_wrench: [ÖbEngine's Development Boards](https://github.com/ObEngine/ObEngine/projects)
- :video_game: [ÖbEngine's Examples](https://github.com/ObEngine/Examples)

## What is ÖbEngine ?
ÖbEngine is a free and open-source 2D game engine ! It focuses on making game development accessible and fun.
ÖbEngine is not only a framework, it is a fully-fledged engine with everything you might need including a map editor, a plugin system and so much more !


*And by the way, it is pronounced* | ˈɒb ˈendʒɪn |, *the umlaut on the Ö is just there because it looks like a surprised face, it doesn't influence the pronounciation.*

## How do I build it ?
- You will need [SFML 2.5.1](https://www.sfml-dev.org/download/sfml/2.5.1/index-fr.php) (Display, Input, Network, Sound and much more)

- You may also need alsa-lib (if you're on Linux)

There are other third-party libraries but they are included in the repository (`extlibs/` folder).
Check this tutorial to learn how to build ÖbEngine : [Building ÖbEngine](https://github.com/ObEngine/ObEngine/wiki/Building-ObEngine)

### Bundled dependencies

#### C++ dependencies
|Library|Description|License|
|-------|-----------|-------|
|[11zip](https://github.com/Sygmei/11Zip)|minizip wrapper based on [Vili Petek's work](http://www.vilipetek.com/2013/11/22/zippingunzipping-files-in-c/)|MIT License|
|[Catch2](https://github.com/catchorg/Catch2)|Test Framework|Boost Software License|
|[dynamicLinker](https://github.com/Marqin/dynamicLinker)|Dynamic Libs wrapper|MIT License|
|[fmt](https://github.com/fmtlib/fmt)|Formatting library|BSD 2-Clause "Simplified" License|
|[Sol](https://github.com/ThePhD/sol)|C++ binding library to Lua|MIT License|
|[Lua 5.3](http://lua-users.org/wiki/LuaBinaries)|Scripting language|zlib License|
|[RichText](https://github.com/Skyrpex/RichText)|sf::Text extension|MIT License|
|[spdlog](https://github.com/gabime/spdlog)|Logging library|MIT License|
|[Vili](https://github.com/Sygmei/Vili)|Homemade Data language (YAML-like)|MIT License|
|[zlib & minizip](http://www.zlib.net/)|zip archives|zlib license|

#### Lua dependencies
|Library|Description|License|
|-------|-----------|-------|
|[debugger.lua](https://github.com/slembcke/debugger.lua)|Embeddable Lua debugger|MIT License|
|[inspect.lua](https://github.com/kikito/inspect.lua)|Human-readable representation of Lua tables|MIT License|

## What I can do with ÖbEngine ?
You can do everything with it as long as it's in 2D. ÖbEngine doesn't handle 3D (maybe partial 3D support in a future update).
You can do some platformers, RPGs, 2D racing games, visual novels, roguelikes, metroidvanias, etc..

If you want to check out games made with ÖbEngine, you can check out those [**open-source projects**](https://github.com/ObEngine/Examples) !

## Is it free ?
Yes it is, completely free ! There is no premium, in-app purchase, forced splashscreen or anything. You can sell your game made with ÖbEngine with no royalties either !
Since the engine is based on the MIT license, you can modify the sources, do anything you want with the engine as long as you keep the original license file somewhere.
We rely exclusively on donations so if you want to support the development you can find the sponsor page [right here](https://github.com/sponsors/Sygmei) !

## On which platforms can I export my game made with ÖbEngine ?

ÖbEngine has been tested on the following platforms :
- Windows XP, 7, 8, 10
- Linux (Debian, Arch)
- MacOS

ÖbEngine will have export for Android, iOS and HTML5 available in a future update.

## Give me some interesting features
Here you go :
- Event-based scripting (with Lua)
- Neat map editor
- Animations
- Native plugins (You can extend the engine with C++ with automatic bindings and documentation generator)
- Canvas (You can draw stuff using a simple API)
- Network support
- Scene / GameObjects system
- Workspaces and Packages system
- Layering system with parallax support
- Polygonal Collisions with full collision detection support
- Integrated CLI to manage your project
- Custom package manager with online repository
- Gamepad support
- Bindable actions with support for complex key combinations
- Shader support
- Audio with many formats supported (OGG, MP3, WAV, FLAC and much more)

## Future features which are also really cool ?
- 3D support
- 2D skeletal animations
- Collaborative map editor
- Light & particle system
- Script the engine with the langage of your choice
- Multiple windows
- Android, iOS and HTML5 export
- Tiled Map Editor support

## ObEngine's versions

Each major release will have a name based on a translation of the word "Eggplant".

ÖbEngine just follows [semver rules](https://semver.org/) which means any version will be tagged (X.Y.Z) with X: Major, Y: Minor and Z: Patch.

|Version number|Version name|Word origin|Release date|Description|
|--------------|------------|-----------|------------|-----------|
|1.0.0|Melanzana|Italian|?|First production release of ÖbEngine|

## Do you have a roadmap I can check ?

Sure ! For now we use GitHub Project system as a planning / roadmap tool which is available [here](https://github.com/ObEngine/ObEngine/projects).

## Right, can I see how does scripting looks ?
Sure, here are some simple GameObjects :
### Hello-World object
This one is really simple, it just prints "Hello World" in the console
```lua
function Local.Init() -- Called when object is created
  print("Hello World");
end
```
Every GameObject can have a Sprite associated (it's cooler when your object appears in the game right ?).
### Rotating goat
Let's imagine you want to create a rotating goat in your game, no problem :
```lua
function Local.Init()
  -- Set the animation for when the goat is flying to the right (You can imagine it already right ?)
  This.Animator:setKey("GOAT_FLYING_LEFT");
end

function Local.Update(dt) -- Local.Update is a function called every loop and dt is the DeltaTime
  This.Sprite:rotate(dt * 45); -- Rotate of 45 degrees each second (You multiply with the DeltaTime here)
end
```

### Drawing stuff
The engine includes a `Canvas` lib to draw stuff in real time and using it is really straightforward !

```lua
function Local.Init()
  local canvas = obe.Canvas.Canvas(400, 400); -- Creating a 400x400 canvas

  canvas:Rectangle("background") { -- Dark grey background
      layer = 2, x = 0, y = 0, width = 250, height = 100,
      color = { r = 50, g = 50, b = 50},
  };

  canvas:Text("firstPlayer") { -- First player's score label
      text = "Player 1 : 0 points", size = 22
  };

  canvas:Text("secondPlayer") { -- Second player's score label
      text = "Player 2 : 0 points", size = 22, y = 50
  };

  canvas:Circle("green") { -- Small green circle
      color = "0F0", -- Green color
      radius = 7, x = 200, y = 5
  };

  canvas:Circle("yellow") { -- Small yellow circle
      color = "FF0", -- Yellow color
      radius = 7, x = 217, y = 5
  };

  canvas:Circle("red") { -- Small red circle
      color = "F00", -- Red color
      radius = 7, x = 234, y = 5
  };

  canvas:render(This.Sprite()); -- Drawing all the stuff !
end
```
Check the Wiki for more examples !

## How can I contribute ?
Interested in contributing to ÖbEngine ? Great ! We always need help on various tasks !

A good first step would be to join the [ÖbEngine's Discord server](https://discord.gg/HkM6zdT) to chat with us and discuss how you could contribute depending on your preferences and skills.


CHANGELOG.md
# Changelog

## 0.4.2

### Minor changes

- Removed useless method `setTarget(target)` from Canvas (replaced with `render(target)`)
- Added default "Hello, world!" project in ÖbEngine, no more crashes when running ÖbEngine for the first time !
- Better exception handling when loading an Animation file
- Added Vili templates for `AnimationPlayMode` enum

### Bugfixes

- Fixed crash when loading a GameObject without Script component

## 0.4.1

### Bugfixes

- Fixed crash when a texture is not found
- Added traceback when a crash happen in Scene's onLoad callback
- Fixed corrupted return values of `getFillColor()` and `getOutlineColor` for Canvas Shapes
- Fixed bug when there was not enough suggestions during an exception
- Fixed invalid value forwarding in vili's node::push
- Fixed bug that ignored the Sprite position defined in .obj.vili
- Fixed bug that swapped GameObject id and type in `NoSuchComponent` exception

## 0.4.0

### Notes

A bit more than two months after 0.3.0 release, here comes the version 0.4 !

There was plans to remove Vili support and move everything to Lua (configuration files, scenes...) but I decided to backpedal and think more about it. In the end, I kept Vili and upgraded it to version 2.0. Here are a few reasons of why I decided to do so :
- Lua can't keep track of the order of insertion
- Exporting Lua data can be tricky
- Unsafe configuration files
- I like Vili :)

You will probably notice a few subtle changes in the Vili files around the engine, the Vili codebase has been massively simplified and this is why some old features are not here anymore (ranges, links, template functions) but these sacrifices are not in vain !

One of this update biggest change is the introduction of a whole brand new exception system, debugging should now be easier. Exceptions now have more details and can even give a few hints on how to fix the problem !
This task took me a lot of time to complete as I had to rewrite all existing exceptions and create new one for the whole ÖbEngine codebase. Exceptions are also handled differently than before, in v0.3.0, when an exception occured in `Event.Game.Update` for example, the GameObject would spam the log with the same exception everytime it happened, now if an exception is raised, the whole engine closes gracefully so you can properly inspect what is going on.

The Canvas also have its share of changes with three new elements : Polygons, Bezier curves and textures. Textures can be applied to all existing shapes of the Canvas.

For this update I was not alone, I got help from [@PierrickLP](https://github.com/PierrickLP) [@Darnagof](https://github.com/Darnagof) and [@Mari0nV](https://github.com/Mari0nV), thanks again !

### Major changes

- Upgraded Vili to version 2.0, new syntax for all the engine configuration / scene files [#231](https://github.com/ObEngine/ObEngine/issues/231) [#232](https://github.com/ObEngine/ObEngine/issues/232) [#263](https://github.com/ObEngine/ObEngine/issues/263) [#276](https://github.com/ObEngine/ObEngine/issues/276) [#277](https://github.com/ObEngine/ObEngine/issues/277) [#278](https://github.com/ObEngine/ObEngine/issues/278)
- Greatly improved exception system with hints and proper traceback [#156](https://github.com/ObEngine/ObEngine/issues/156)
- Standardized time unit, every time value in the engine is now as seconds (double precision float) [#255](https://github.com/ObEngine/ObEngine/issues/255)
- Added Bezier curves, Polygons and Sprites (as Rect textures) support to the Canvas [#161](https://github.com/ObEngine/ObEngine/issues/161)
- Engine now properly closes when an exception is encountered [#226](https://github.com/ObEngine/ObEngine/issues/226)

### Minor changes

- Possibility to import lua files from mounted paths with `require` function (thanks to [@Mari0nV](https://github.com/Mari0nV)) [#275](https://github.com/ObEngine/ObEngine/issues/275)
- All Sprite attributes can now be modified from configuration file (thanks to [@PierrickLP](https://github.com/PierrickLP)) [#270](https://github.com/ObEngine/ObEngine/issues/270) [#273](https://github.com/ObEngine/ObEngine/issues/273)
- Better collision checks results (thanks to [@Darnagof](https://github.com/Darnagof)) [#271](https://github.com/ObEngine/ObEngine/issues/271)
- It is now possible to remove a child from a SceneNode (suggestion of [@PierrickLP](https://github.com/PierrickLP)) [#256](https://github.com/ObEngine/ObEngine/issues/256)
- More flexibility when editing Canvas.Text attributes [#197](https://github.com/ObEngine/ObEngine/issues/197)
- Improvements to the the Color class [#249](https://github.com/ObEngine/ObEngine/issues/249) [#261](https://github.com/ObEngine/ObEngine/issues/261)
- Logger now available from Lua [#280](https://github.com/ObEngine/ObEngine/issues/280)

src/Core/Graphics/Color.cpp
#include "Utils/MathUtils.hpp"

#include <Graphics/Color.hpp>
#include <Graphics/Exceptions.hpp>

#include <array>
#include <cmath>
#include <sstream>
#include <unordered_map>

namespace obe::Graphics
{
    Color Color::AliceBlue(240, 248, 255);
    Color Color::AntiqueWhite(250, 235, 215);
    Color Color::Aqua(0, 255, 255);
    Color Color::Aquamarine(127, 255, 212);
    Color Color::Azure(240, 255, 255);
    Color Color::Beige(245, 245, 220);
    Color Color::Bisque(255, 228, 196);
    Color Color::Black(0, 0, 0);
    Color Color::BlanchedAlmond(255, 235, 205);
    Color Color::Blue(0, 0, 255);
    Color Color::BlueViolet(138, 43, 226);
    Color Color::Brown(165, 42, 42);
    Color Color::BurlyWood(222, 184, 135);
    Color Color::CadetBlue(95, 158, 160);
    Color Color::Chartreuse(127, 255, 0);
    Color Color::Chocolate(210, 105, 30);
    Color Color::Coral(255, 127, 80);
    Color Color::CornflowerBlue(100, 149, 237);
    Color Color::Cornsilk(255, 248, 220);
    Color Color::Crimson(220, 20, 60);
    Color Color::Cyan(0, 255, 255);
    Color Color::DarkBlue(0, 0, 139);
    Color Color::DarkCyan(0, 139, 139);
    Color Color::DarkGoldenrod(184, 134, 11);
    Color Color::DarkGray(169, 169, 169);
    Color Color::DarkGreen(0, 100, 0);
    Color Color::DarkKhaki(189, 183, 107);
    Color Color::DarkMagenta(139, 0, 139);
    Color Color::DarkOliveGreen(85, 107, 47);
    Color Color::DarkOrange(255, 140, 0);
    Color Color::DarkOrchid(153, 50, 204);
    Color Color::DarkRed(139, 0, 0);
    Color Color::DarkSalmon(233, 150, 122);
    Color Color::DarkSeaGreen(143, 188, 139);
    Color Color::DarkSlateBlue(72, 61, 139);
    Color Color::DarkSlateGray(47, 79, 79);
    Color Color::DarkTurquoise(0, 206, 209);
    Color Color::DarkViolet(148, 0, 211);
    Color Color::DeepPink(255, 20, 147);
    Color Color::DeepSkyBlue(0, 191, 255);
    Color Color::DimGray(105, 105, 105);
    Color Color::DodgerBlue(30, 144, 255);
    Color Color::FireBrick(178, 34, 34);
    Color Color::FloralWhite(255, 250, 240);
    Color Color::ForestGreen(34, 139, 34);
    Color Color::Fuchsia(255, 0, 255);
    Color Color::Gainsboro(220, 220, 220);
    Color Color::GhostWhite(248, 248, 255);
    Color Color::Gold(255, 215, 0);
    Color Color::Goldenrod(218, 165, 32);
    Color Color::Gray(128, 128, 128);
    Color Color::Green(0, 128, 0);
    Color Color::GreenYellow(173, 255, 47);
    Color Color::HoneyDew(240, 255, 240);
    Color Color::HotPink(255, 105, 180);
    Color Color::IndianRed(205, 92, 92);
    Color Color::Indigo(75, 0, 130);
    Color Color::Ivory(255, 255, 240);
    Color Color::Khaki(240, 230, 140);
    Color Color::Lavender(230, 230, 250);
    Color Color::LavenderBlush(255, 240, 245);
    Color Color::LawnGreen(124, 252, 0);
    Color Color::LemonChiffon(255, 250, 205);
    Color Color::LightBlue(173, 216, 230);
    Color Color::LightCoral(240, 128, 128);
    Color Color::LightCyan(224, 255, 255);
    Color Color::LightGoldenrodYellow(250, 250, 210);
    Color Color::LightGray(211, 211, 211);
    Color Color::LightGreen(144, 238, 144);
    Color Color::LightPink(255, 182, 193);
    Color Color::LightSalmon(255, 160, 122);
    Color Color::LightSeaGreen(32, 178, 170);
    Color Color::LightSkyBlue(135, 206, 250);
    Color Color::LightSlateGray(119, 136, 153);
    Color Color::LightSteelBlue(176, 196, 222);
    Color Color::LightYellow(255, 255, 224);
    Color Color::Lime(0, 255, 0);
    Color Color::LimeGreen(50, 205, 50);
    Color Color::Linen(250, 240, 230);
    Color Color::Magenta(255, 0, 255);
    Color Color::Maroon(128, 0, 0);
    Color Color::MediumAquamarine(102, 205, 170);
    Color Color::MediumBlue(0, 0, 205);
    Color Color::MediumOrchid(186, 85, 211);
    Color Color::MediumPurple(147, 112, 219);
    Color Color::MediumSeaGreen(60, 179, 113);
    Color Color::MediumSlateBlue(123, 104, 238);
    Color Color::MediumSpringGreen(0, 250, 154);
    Color Color::MediumTurquoise(72, 209, 204);
    Color Color::MediumVioletRed(199, 21, 133);
    Color Color::MidnightBlue(25, 25, 112);
    Color Color::MintCream(245, 255, 250);
    Color Color::MistyRose(255, 228, 225);
    Color Color::Moccasin(255, 228, 181);
    Color Color::NavajoWhite(255, 222, 173);
    Color Color::Navy(0, 0, 128);
    Color Color::OldLace(253, 245, 230);
    Color Color::Olive(128, 128, 0);
    Color Color::OliveDrab(107, 142, 35);
    Color Color::Orange(255, 165, 0);
    Color Color::OrangeRed(255, 69, 0);
    Color Color::Orchid(218, 112, 214);
    Color Color::PaleGoldenrod(238, 232, 170);
    Color Color::PaleGreen(152, 251, 152);
    Color Color::PaleTurquoise(175, 238, 238);
    Color Color::PaleVioletRed(219, 112, 147);
    Color Color::PapayaWhip(255, 239, 213);
    Color Color::PeachPuff(255, 218, 185);
    Color Color::Peru(205, 133, 63);
    Color Color::Pink(255, 192, 203);
    Color Color::Plum(221, 160, 221);
    Color Color::PowderBlue(176, 224, 230);
    Color Color::Purple(128, 0, 128);
    Color Color::RebeccaPurple(102, 51, 153);
    Color Color::Red(255, 0, 0);
    Color Color::RosyBrown(188, 143, 143);
    Color Color::RoyalBlue(65, 105, 225);
    Color Color::SaddleBrown(139, 69, 19);
    Color Color::Salmon(250, 128, 114);
    Color Color::SandyBrown(244, 164, 96);
    Color Color::SeaGreen(46, 139, 87);
    Color Color::SeaShell(255, 245, 238);
    Color Color::Sienna(160, 82, 45);
    Color Color::Silver(192, 192, 192);
    Color Color::SkyBlue(135, 206, 235);
    Color Color::SlateBlue(106, 90, 205);
    Color Color::SlateGray(112, 128, 144);
    Color Color::Snow(255, 250, 250);
    Color Color::SpringGreen(0, 255, 127);
    Color Color::SteelBlue(70, 130, 180);
    Color Color::Tan(210, 180, 140);
    Color Color::Teal(0, 128, 128);
    Color Color::Thistle(216, 191, 216);
    Color Color::Tomato(255, 99, 71);
    Color Color::Turquoise(64, 224, 208);
    Color Color::Violet(238, 130, 238);
    Color Color::Wheat(245, 222, 179);
    Color Color::White(255, 255, 255);
    Color Color::WhiteSmoke(245, 245, 245);
    Color Color::Yellow(255, 255, 0);
    Color Color::YellowGreen(154, 205, 50);

    static std::unordered_map<std::string, Color> ColorNames = { { "aliceblue",
                                                              Color::AliceBlue },
        { "antiquewhite", Color::AntiqueWhite }, { "aqua", Color::Aqua },
        { "aquamarine", Color::Aquamarine }, { "azure", Color::Azure },
        { "beige", Color::Beige }, { "bisque", Color::Bisque }, { "black", Color::Black },
        { "blanchedalmond", Color::BlanchedAlmond }, { "blue", Color::Blue },
        { "blueviolet", Color::BlueViolet }, { "brown", Color::Brown },
        { "burlywood", Color::BurlyWood }, { "cadetblue", Color::CadetBlue },
        { "chartreuse", Color::Chartreuse }, { "chocolate", Color::Chocolate },
        { "coral", Color::Coral }, { "cornflowerblue", Color::CornflowerBlue },
        { "cornsilk", Color::Cornsilk }, { "crimson", Color::Crimson },
        { "cyan", Color::Cyan }, { "darkblue", Color::DarkBlue },
        { "darkcyan", Color::DarkCyan }, { "darkgoldenrod", Color::DarkGoldenrod },
        { "darkgray", Color::DarkGray }, { "darkgreen", Color::DarkGreen },
        { "darkkhaki", Color::DarkKhaki }, { "darkmagenta", Color::DarkMagenta },
        { "darkolivegreen", Color::DarkOliveGreen }, { "darkorange", Color::DarkOrange },
        { "darkorchid", Color::DarkOrchid }, { "darkred", Color::DarkRed },
        { "darksalmon", Color::DarkSalmon }, { "darkseagreen", Color::DarkSeaGreen },
        { "darkslateblue", Color::DarkSlateBlue },
        { "darkslategray", Color::DarkSlateGray },
        { "darkturquoise", Color::DarkTurquoise }, { "darkviolet", Color::DarkViolet },
        { "deeppink", Color::DeepPink }, { "deepskyblue", Color::DeepSkyBlue },
        { "dimgray", Color::DimGray }, { "dodgerblue", Color::DodgerBlue },
        { "firebrick", Color::FireBrick }, { "floralwhite", Color::FloralWhite },
        { "forestgreen", Color::ForestGreen }, { "fuchsia", Color::Fuchsia },
        { "gainsboro", Color::Gainsboro }, { "ghostwhite", Color::GhostWhite },
        { "gold", Color::Gold }, { "goldenrod", Color::Goldenrod },
        { "gray", Color::Gray }, { "green", Color::Green },
        { "greenyellow", Color::GreenYellow }, { "honeydew", Color::HoneyDew },
        { "hotpink", Color::HotPink }, { "indianred", Color::IndianRed },
        { "indigo", Color::Indigo }, { "ivory", Color::Ivory }, { "khaki", Color::Khaki },
        { "lavender", Color::Lavender }, { "lavenderblush", Color::LavenderBlush },
        { "lawngreen", Color::LawnGreen }, { "lemonchiffon", Color::LemonChiffon },
        { "lightblue", Color::LightBlue }, { "lightcoral", Color::LightCoral },
        { "lightcyan", Color::LightCyan },
        { "lightgoldenrodyellow", Color::LightGoldenrodYellow },
        { "lightgray", Color::LightGray }, { "lightgreen", Color::LightGreen },
        { "lightpink", Color::LightPink }, { "lightsalmon", Color::LightSalmon },
        { "lightsalmon", Color::LightSalmon }, { "lightseagreen", Color::LightSeaGreen },
        { "lightskyblue", Color::LightSkyBlue },
        { "lightslategray", Color::LightSlateGray },
        { "lightsteelblue", Color::LightSteelBlue },
        { "lightyellow", Color::LightYellow }, { "lime", Color::Lime },
        { "limegreen", Color::LimeGreen }, { "linen", Color::Linen },
        { "magenta", Color::Magenta }, { "maroon", Color::Maroon },
        { "mediumaquamarine", Color::MediumAquamarine },
        { "mediumblue", Color::MediumBlue }, { "mediumorchid", Color::MediumOrchid },
        { "mediumpurple", Color::MediumPurple },
        { "mediumseagreen", Color::MediumSeaGreen },
        { "mediumslateblue", Color::MediumSlateBlue },
        { "mediumslateblue", Color::MediumSlateBlue },
        { "mediumspringgreen", Color::MediumSpringGreen },
        { "mediumturquoise", Color::MediumTurquoise },
        { "mediumvioletred", Color::MediumVioletRed },
        { "midnightblue", Color::MidnightBlue }, { "mintcream", Color::MintCream },
        { "mistyrose", Color::MistyRose }, { "moccasin", Color::Moccasin },
        { "navajowhite", Color::NavajoWhite }, { "navy", Color::Navy },
        { "oldlace", Color::OldLace }, { "olive", Color::Olive },
        { "olivedrab", Color::OliveDrab }, { "orange", Color::Orange },
        { "orangered", Color::OrangeRed }, { "orchid", Color::Orchid },
        { "palegoldenrod", Color::PaleGoldenrod }, { "palegreen", Color::PaleGreen },
        { "paleturquoise", Color::PaleTurquoise },
        { "palevioletred", Color::PaleVioletRed }, { "papayawhip", Color::PapayaWhip },
        { "peachpuff", Color::PeachPuff }, { "peru", Color::Peru },
        { "pink", Color::Pink }, { "plum", Color::Plum },
        { "powderblue", Color::PowderBlue }, { "purple", Color::Purple },
        { "rebeccapurple", Color::RebeccaPurple }, { "red", Color::Red },
        { "rosybrown", Color::RosyBrown }, { "royalblue", Color::RoyalBlue },
        { "saddlebrown", Color::SaddleBrown }, { "salmon", Color::Salmon },
        { "sandybrown", Color::SandyBrown }, { "seagreen", Color::SeaGreen },
        { "seashell", Color::SeaShell }, { "sienna", Color::Sienna },
        { "silver", Color::Silver }, { "skyblue", Color::SkyBlue },
        { "slateblue", Color::SlateBlue }, { "slategray", Color::SlateGray },
        { "snow", Color::Snow }, { "springgreen", Color::SpringGreen },
        { "steelblue", Color::SteelBlue }, { "tan", Color::Tan }, { "teal", Color::Teal },
        { "thistle", Color::Thistle }, { "tomato", Color::Tomato },
        { "turquoise", Color::Turquoise }, { "violet", Color::Violet },
        { "wheat", Color::Wheat }, { "white", Color::White },
        { "whitesmoke", Color::WhiteSmoke }, { "yellow", Color::Yellow },
        { "yellowgreen", Color::YellowGreen } };

    Color::Color()
    {
    }

    Color::Color(const double r, const double g, const double b, const double a)
    {
        this->fromRgb(r, g, b, a);
    }

    Color::Color(const std::string& nameOrHex)
    {
        this->fromString(nameOrHex);
    }


    Color::Color(const sf::Color& color)
    {
        this->r = color.r;
        this->g = color.g;
        this->b = color.b;
        this->a = color.a;
    }

    void Color::fromString(std::string string)
    {
        if (!this->fromName(string, false))
        {
            this->fromHex(string);
        }
    }

    bool Color::fromName(std::string name, bool strict)
    {
        std::for_each(name.begin(), name.end(), [](char& c) { c = std::tolower(c); });
        if (const auto& color = ColorNames.find(name); color != ColorNames.end())
        {
            this->r = color->second.r;
            this->g = color->second.g;
            this->b = color->second.b;
            this->a = color->second.a;
            return true;
        }
        if (!strict)
            return false;
        throw Exceptions::InvalidColorName(name, EXC_INFO);
    }

    void Color::fromRgb(const double r, const double g, const double b, const double a)
    {
        if (r < 0 || r > 255 || g < 0 || g > 255 || b < 0 || b > 255 || a < 0 || a > 255)
        {
            throw Exceptions::InvalidRgbFormat(r, g, b, a, EXC_INFO);
        }
        this->r = r;
        this->g = g;
        this->b = b;
        this->a = a;
    }

    void Color::fromHsv(const int H, const double S, const double V)
    {

        if (H < 0 || H > 360 || S < 0.0 || S > 1.0 || V < 0.0 || V > 1.0)
        {
            throw Exceptions::InvalidHsvFormat(H, S, V, EXC_INFO);
        }

        const double C = S * V;
        const double X = C * (1 - abs(fmod(H / 60.0, 2) - 1));
        const double m = V - C;
        double Rs, Gs, Bs;

        if (H >= 0 && H < 60)
        {
            Rs = C;
            Gs = X;
            Bs = 0;
        }
        else if (H >= 60 && H < 120)
        {
            Rs = X;
            Gs = C;
            Bs = 0;
        }
        else if (H >= 120 && H < 180)
        {
            Rs = 0;
            Gs = C;
            Bs = X;
        }
        else if (H >= 180 && H < 240)
        {
            Rs = 0;
            Gs = X;
            Bs = C;
        }
        else if (H >= 240 && H < 300)
        {
            Rs = X;
            Gs = 0;
            Bs = C;
        }
        else
        {
            Rs = C;
            Gs = 0;
            Bs = X;
        }

        this->r = (Rs + m) * 255;
        this->g = (Gs + m) * 255;
        this->b = (Bs + m) * 255;
    }

    uint32_t Color::toInteger() const
    {
        return (static_cast<int>(r) << 24) + (static_cast<int>(g) << 16)
            + (static_cast<int>(b) << 8) + (a);
    }

    std::string Color::toHex() const
    {
        std::stringstream stream;
        stream << "#" << std::hex << r << g << b;
        if (a != 1)
        {
            stream << a;
        }
        return stream.str();
    }

    std::string Color::toName() const
    {
        for (const auto& color : ColorNames)
        {
            if (color.second.r == r && color.second.g == g && color.second.b == b)
            {
                return color.first;
            }
        }

        return "";
    }

    void Color::fromHex(std::string hexCode)
    {
        std::array<unsigned short, 3> rgb {};
        std::stringstream ss;
        std::string str;

        if (hexCode[0] == '#')
        {
            hexCode.erase(0, 1);
        }

        const int size = hexCode.size();

        if (size != 3 && size != 6
            || hexCode.find_first_not_of("AaBbCcDdEeFf0123456789") != std::string::npos)
        {
            throw Exceptions::InvalidHexFormat(hexCode, EXC_INFO);
        }

        for (unsigned int i = 0; i < 3; i++)
        {
            if (size == 3)
            {
                str = std::string(2, hexCode[i]);
            }
            else if (size == 6)
            {
                str = hexCode.substr(i * 2, 2);
            }
            else
            {
                break;
            }
            ss << std::hex << str;
            ss >> rgb[i];
            ss.clear();
        }

        this->r = rgb[0];
        this->g = rgb[1];
        this->b = rgb[2];
    }

    bool Color::operator==(const Color& color) const
    {
        return toInteger() == color.toInteger();
    }

    bool Color::operator!=(const Color& color) const
    {
        return toInteger() != color.toInteger();
    }

    Color Color::operator+(const Color& color) const
    {

        return Color(r + color.r, g + color.g, b + color.b, a + color.a);
    }

    void Color::operator+=(const Color& color)
    {

        r += color.r;
        g += color.g;
        b += color.b;
        a += color.a;
    }

    Color Color::operator-(const Color& color) const
    {

        return Color(r - color.r, g - color.g, b - color.b, a - color.a);
    }

    void Color::operator-=(const Color& color)
    {

        r -= color.r;
        g -= color.g;
        b -= color.b;
        a -= color.a;
    }

    Color Color::operator*(const Color& color) const
    {

        return Color(r * color.r, g * color.g, b * color.b, a * color.a);
    }

    Color Color::operator*(double value) const
    {

        return Color(r * value, g * value, b * value, a * value);
    }

    void Color::operator*=(const Color& color)
    {

        r = r * color.r;
        g = g * color.g;
        b = b * color.b;
        a = a * color.a;
    }

    void Color::operator*=(double value)
    {

        r = r * value;
        g = g * value;
        b = b * value;
        a = a * value;
    }

    Color Color::operator/(const Color& color) const
    {

        return Color(r / color.r, g / color.g, b / color.b, a / color.a);
    }

    Color Color::operator/(double value) const
    {

        return Color(r / value, g / value, b / value, a / value);
    }

    void Color::operator/=(const Color& color)
    {

        r = r / color.r;
        g = g / color.g;
        b = b / color.b;
        a = a / color.a;
    }

    void Color::operator/=(double value)
    {

        if (value != 0)
        {
            r = r / value;
            g = g / value;
            b = b / value;
            a = a / value;
        }
        else
        {
            r = 255;
            g = 255;
            b = 255;
            a = 255;
        }
    };

    Color Color::operator-() const
    {

        return Color(255 - r, 255 - g, 255 - b, 255 - a);
    }

    Color::operator sf::Color() const
    {
        return sf::Color(r, g, b, a);
    }

    Color Color::Random(bool randomAlpha)
    {
        int alpha = 255;
        if (randomAlpha)
        {
            alpha = Utils::Math::randfloat() * 255.f;
        }
        return Color(255.f * Utils::Math::randfloat(), 255.f * Utils::Math::randfloat(),
            255.f * Utils::Math::randfloat(), alpha);
    }

    std::ostream& operator<<(std::ostream& os, const Color& color)
    {
        os << "Color(" << int(color.r) << ", " << int(color.g) << ", " << int(color.b)
           << ", " << int(color.a) << ")";
        return os;
    }
} // namespace obe::Graphics
